File: src/com/dafttech/terra/engine/gui/MouseSlot.java
package com.dafttech.terra.engine.gui;

import com.badlogic.gdx.graphics.Color;
import com.dafttech.eventmanager.Event;
import com.dafttech.eventmanager.EventListener;
import com.dafttech.terra.engine.AbstractScreen;
import com.dafttech.terra.engine.Vector2;
import com.dafttech.terra.engine.gui.elements.ElementSlot;
import com.dafttech.terra.game.Events;
import com.dafttech.terra.game.world.items.Item;
import com.dafttech.terra.game.world.items.persistence.Prototype;
import com.dafttech.terra.resources.Resources;

public class MouseSlot {       
    private static class MouseRenderSlot extends ElementSlot {
        public MouseRenderSlot(Vector2 p) {
            super(p);
            Events.EVENTMANAGER.registerEventListener(this);
        }
        
        @EventListener("MOUSEMOVE")
        public void onEventMouseMove(Event event) {
            int x = (int) event.getInput()[1];
            int y = (int) event.getInput()[2];

            position.x = x;
            position.y = y;
        }
        
        @Override
        public void draw(AbstractScreen screen) {
            Vector2 p = getScreenPosition();

            screen.batch.begin();

            if (assignedType != null) {
                ((Item) assignedType.toGameObject()).drawInventory(p, screen);
            }

            screen.batch.end();
        }
    }
    
    static private Prototype assignedType = null;
    static private MouseRenderSlot renderSlot = new MouseRenderSlot(new Vector2(0, 0));
    
    public static void init() {
    }
    
    public static ElementSlot getRenderSlot() {
        return renderSlot;
    }
    
    public static boolean canAssignType() {
        return assignedType == null;
    }
    
    public static Prototype getAssignedType() {
        return assignedType;
    }
    
    public static Prototype popAssignedType() {
        Prototype at = assignedType;
        assignedType = null;
        renderSlot.assignType(null);
        return at;
    }
    
    public static boolean assignType(Prototype type) {
        if (assignedType == null) {
            assignedType = type;
            renderSlot.assignType(assignedType);
            return true;
        }
        return false;
    }
}

File: src/main/java/com/dafttech/terra/engine/gui/MouseSlot.scala
package com.dafttech.terra.engine.gui

import com.dafttech.terra.engine.gui.elements.ElementSlot
import com.dafttech.terra.engine.{AbstractScreen, Vector2}
import com.dafttech.terra.game.Events
import com.dafttech.terra.game.world.items.Item
import com.dafttech.terra.game.world.items.inventories.Stack
import org.lolhens.eventmanager.{Event, EventListener}

object MouseSlot {

  private class MouseRenderSlot(p: Vector2) extends ElementSlot(p) {
    Events.EVENTMANAGER.registerEventListener(this)

    @EventListener(Array("MOUSEMOVE"))
    def onEventMouseMove(event: Event) {
      val x: Int = event.in.get(1, classOf[Integer])
      val y: Int = event.in.get(2, classOf[Integer])
      position.x = x
      position.y = y
    }

    override def draw(screen: AbstractScreen) {
      val p: Vector2 = getScreenPosition
      screen.batch.begin
      if (assignedStack != null) {
        (assignedStack.`type`.toGameObject.asInstanceOf[Item]).drawInventory(p, screen)
      }
      screen.batch.end
    }
  }

  private var assignedStack: Stack = null
  private var renderSlot: MouseSlot.MouseRenderSlot = new MouseSlot.MouseRenderSlot(new Vector2(0, 0))

  def init {
  }

  def getRenderSlot: ElementSlot = {
    return renderSlot
  }

  def canAssignStack: Boolean = {
    return assignedStack == null
  }

  def getAssignedStack: Stack = {
    return assignedStack
  }

  def popAssignedStack: Stack = {
    val at: Stack = assignedStack
    assignedStack = null
    renderSlot.assignStack(null)
    return at
  }

  def assignStack(stack: Stack): Boolean = {
    if (assignedStack == null) {
      assignedStack = stack
      renderSlot.assignStack(assignedStack)
      return true
    }
    return false
  }
}